[PrimTorch] Add maker for *_copy variants of view functions

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87182
* #87183
* #87205
* #87204
* #87203
* #87424
* #87181
* #87282
* #87180
* __->__ #87278
* #87202

Implements `diagonal_copy` as an example. This PR also fixes a number of
correcness issues with `diagonal_copy`.

cc @ezyang @mruberry @ngimel @Lezcano @peterbell10 @fdrocha

lezcano committed:
[PrimTorch] Add maker for *_copy variants of view functions

Implements `diagonal_copy` as an example. This PR also fixes a number of
correcness issues with `diagonal_copy`.

[ghstack-poisoned]

lezcano commented:
@bdhirsh I had to hardcode many things here so that the op passes the tests. These could be easily codegen'd. Same in the out version codegen'd from `diag_embed`. See
https://github.com/pytorch/pytorch/blob/3531f98c742612b8a604b459509762c495064436/aten/src/ATen/native/TensorShape.cpp#L3408-L3412

lezcano committed:
Update on "[PrimTorch] Add maker for *_copy variants of view functions"


Implements `diagonal_copy` as an example. This PR also fixes a number of
correcness issues with `diagonal_copy`.

[ghstack-poisoned]

lezcano committed:
Update on "[PrimTorch] Add maker for *_copy variants of view functions"


Implements `diagonal_copy` as an example. This PR also fixes a number of
correcness issues with `diagonal_copy`.

[ghstack-poisoned]

lezcano commented on test/functorch/test_aotdispatch.py:
```diff
@@ -990,6 +990,7 @@ def assert_compiler(gm: torch.fx.GraphModule, _):
     xfail('deg2rad', ''),  # aten.deg2rad.default - couldn't find symbolic meta function/decomposition
     xfail('diag', ''),  # Cannot call sizes() on tensor with symbolic sizes/strides
     xfail('diagonal', ''),  # Cannot call sizes() on tensor with symbolic sizes/strides
+    xfail('diagonal_copy', ''),  # Cannot call sizes() on tensor with symbolic sizes/strides
     xfail('diagonal_scatter', ''),  # Cannot call sizes() on tensor with symbolic sizes/strides
```
I'm fixing these at a later PR adding a decomposition for `diagonal_scatter`.

lezcano committed:
Update on "[PrimTorch] Add maker for *_copy variants of view functions"


Implements `diagonal_copy` as an example. This PR also fixes a number of
correcness issues with `diagonal_copy`.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

albanD reviewed (commented):
Note that this will conflict with some of the stuff we already have in the symbolic shape branch.

lezcano committed:
Update on "[PrimTorch] Add maker for *_copy variants of view functions"


Implements `diagonal_copy` as an example. This PR also fixes a number of
correcness issues with `diagonal_copy`.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -1972,6 +1973,26 @@ def _reduction(
     return result
 
 
+def _make_copy_from_view(fn):
```
Add a description for this helper

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -1972,6 +1973,26 @@ def _reduction(
+def _make_copy_from_view(fn):
+    # Does not handle multiple outputs, but I think we don't have any such an op
+    name = fn.__name__
+    fn = out_wrapper()(fn)
```
Let's be very clear about why this applies the out wrapper instead of asking the developer to do it

lezcano replied:
In fact, I believe that this application of `out_wrapper` should not be necessary. I should delete it and simply reimplement the logic, as I've done below.

mruberry replied:
(offline discussion between @mruberry and @lezcano suggested a tweak to this idea)

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -1972,6 +1973,26 @@ def _reduction(
+            return out
+
+    copy_name = f"{name}_copy"
+    _fn.__name__ = copy_name
```
Note for our future selves: we should probably create a standard for wrappers like this. Here we're updating the name but `out_wrapper` is more aggressive:

https://github.com/pytorch/pytorch/blob/75a5a46aa005e1c56f5a7935003cb480f33f9257/torch/_prims_common/wrappers.py#L264

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5134,7 +5134,7 @@ def reference_inputs_diagonal_diag_embed(op_info, device, dtype, requires_grad,
 
     for shape, kwargs in chain(samples1d, samples2d, samples3d):
-        if op_info.name in ('diagonal', '_refs.diagonal'):
+        if 'diagonal' in op_info.name:
```
Nit: I don't like this name checks and would prefer we test for whether functions are equivalent instead:

```
if op.op in (torch.diagonal, ...)
```

mruberry replied:
(Even better is to control this behavior through a kwarg on the function and have the OpInfos `partial` their sample inputs with the kwarg to control the behavior)

lezcano replied:
I put this because otherwise I would have needed to put 4 different ops, which is a bit tiresome and a error-prone.

mruberry reviewed (approved):
Approving for velocity -- necessary changes seem straightforward

lezcano committed:
Update on "[PrimTorch] Add maker for *_copy variants of view functions"


Implements `diagonal_copy` as an example. This PR also fixes a number of
correcness issues with `diagonal_copy`.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

lezcano committed:
Update on "[PrimTorch] Add maker for *_copy variants of view functions"


Implements `diagonal_copy` as an example. This PR also fixes a number of
correcness issues with `diagonal_copy`.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

lezcano committed:
Update on "[PrimTorch] Add maker for *_copy variants of view functions"


Implements `diagonal_copy` as an example. This PR also fixes a number of
correcness issues with `diagonal_copy`.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

lezcano committed:
Update on "[PrimTorch] Add maker for *_copy variants of view functions"


Implements `diagonal_copy` as an example. This PR also fixes a number of
correcness issues with `diagonal_copy`.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]

lezcano commented on torch/_refs/__init__.py:
```diff
@@ -1972,6 +1973,26 @@ def _reduction(
+    def _fn(*args, out=None, **kwargs):
+        result = fn(*args, **kwargs)
+        if out is None:
+            return result.clone()
```
updated the logic!

lezcano committed:
Update on "[PrimTorch] Add maker for *_copy variants of view functions"


Implements `diagonal_copy` as an example. This PR also fixes a number of
correcness issues with `diagonal_copy`.

cc ezyang mruberry ngimel Lezcano fdrocha

[ghstack-poisoned]
